src/biogeophys/SnowHydrologyMod.F90

```diff
@@ -2279,6 +2292,97 @@ subroutine InitSnowLayers (bounds, snow_depth)
          zi  => col%zi     & ! Output: [real(r8) (:,:) ]  interface level below a "z" level (m) (-nlevsno+0:nlevgrnd)
     )
 
+    allocate(dzmin(1:nlevsno))
+    allocate(dzmax_l(1:nlevsno))
+    allocate(dzmax_u(1:nlevsno))
+
+    ! FOR TEMPORARY TESTING. DELETE WHEN DONE. slevis diag
+    allocate(dzmin_orig(1:nlevsno))
+    allocate(dzmax_l_orig(1:nlevsno))
+    allocate(dzmax_u_orig(1:nlevsno))
+
+    ! These three variables determine the vertical structure of the snow pack:
+    ! dzmin: minimum snow thickness of layer
+    ! dzmax_l: maximum snow thickness of layer when no layers beneath
+    ! dzmax_u: maximum snow thickness of layer when layers beneath
+    dzmin(1) = snow_dzmin_1  ! default or user-defined value from namelist
+    dzmax_l(1) = snow_dzmax_l_1  ! same comment
+    dzmax_u(1) = snow_dzmax_u_1  ! same comment
+    dzmin(2) = snow_dzmin_2  ! default or user-defined value from namelist
+    dzmax_l(2) = snow_dzmax_l_2  ! same comment
+    dzmax_u(2) = snow_dzmax_u_2  ! same comment
+    do j = 3, nlevsno
+       dzmin(j) = dzmax_u(j-1) * 0.5_r8
+       dzmax_u(j) = 2._r8 * dzmax_u(j-1) + 0.01_r8
+       dzmax_l(j) = dzmax_u(j) + dzmax_l(j-1)
+       ! error checks
+       if (dzmin(j) <= dzmin(j-1)) then
+          write(iulog,*) 'ERROR at snow layer j =', j, ' because dzmin(j) =', dzmin(j), ' and dzmin(j-1) =', dzmin(j-1)
+          call endrun(msg="ERROR dzmin(j) cannot be <= dzmin(j-1)"// &
+               errMsg(sourcefile, __LINE__))
+       end if
+       if (dzmax_u(j) <= dzmax_u(j-1)) then
+          write(iulog,*) 'ERROR at snow layer j =', j, ' because dzmax_u(j) =', dzmax_u(j), ' and dzmax_u(j-1) =', dzmax_u(j-1)
+          call endrun(msg="ERROR dzmax_u(j) cannot be <= dzmax_u(j-1)"// &
+               errMsg(sourcefile, __LINE__))
+       end if
+       if (dzmax_l(j) <= dzmax_l(j-1)) then
+          write(iulog,*) 'ERROR at snow layer j =', j, ' because dzmax_l(j) =', dzmax_l(j), ' and dzmax_l(j-1) =', dzmax_l(j-1)
+          call endrun(msg="ERROR dzmax_l(j) cannot be <= dzmax_l(j-1)"// &
+               errMsg(sourcefile, __LINE__))
+       end if
+       if (dzmin(j) >= dzmax_u(j)) then
+          write(iulog,*) 'ERROR at snow layer j =', j, ' because dzmin(j) =', dzmin(j), ' and dzmax_u(j) =', dzmax_u(j)
+          call endrun(msg="ERROR dzmin(j) cannot be >= dzmax_u(j)"// &
+               errMsg(sourcefile, __LINE__))
+       end if
+       if (dzmax_u(j) >= dzmax_l(j)) then
+          write(iulog,*) 'ERROR at snow layer j =', j, ' because dzmax_u(j) =', dzmax_u(j), ' and dzmax_l(j) =', dzmax_l(j)
+          call endrun(msg="ERROR dzmax_u(j) cannot be >= dzmax_l(j)"// &
+               errMsg(sourcefile, __LINE__))
+       end if
+    end do
+    if (nlevsno >= 12) then  ! max nlevsno is 12
```

Comment:
Initially this did not make sense to me because I had not thought it through. But now I think I understand: is the idea that even with a maximum of 3 snow layers, the 3rd needs to be unlimited to allow for deep snow?

Reply:
Done.

Reply:
Yes, that's right - thanks. Sorry that I didn't explain my thinking clearly earlier.

The adjusted code looks good for this, but as a very minor point, I'd remove the comment, "The original code set layer 12 to huge...": comments like this quickly lose their value, because it's unclear what "original code" this is referring to and why. (Comments like this are useful in the context of a PR, e.g., as line comments in the PR, but are unnecessary in the final code.)